RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc

```diff
@@ -0,0 +1,413 @@
+    mat[2][2] = 0.2 * 0.2;    // sigma^2(phi)
+    mat[3][3] = 20. * 20.;    // sigma^2(x_transverse))
+    mat[4][4] = 20. * 20.;    // sigma^2(y_transverse))
+
+    CurvilinearTrajectoryError error(mat);
+
+    const FreeTrajectoryState state(param, error);
+
+    LogDebug(metname) << "Free trajectory State from the parameters";
+    LogDebug(metname) << debug.dumpFTS(state);
+
+    // Propagate the state on the MB2/ME2 surface
+    TrajectoryStateOnSurface tsos = theService->propagator(thePropagatorName)->propagate(state, detLayer->surface());
+
+    LogDebug(metname) << "State after the propagation on the layer";
+    LogDebug(metname) << debug.dumpLayer(detLayer);
+    LogDebug(metname) << debug.dumpTSOS(tsos);
+
+    double dRcone = matchingDR[0];
+    if (fabs(eta) < etaBins.back()) {
+      std::vector<double>::iterator lowEdge = std::upper_bound(etaBins.begin(), etaBins.end(), fabs(eta));
+      dRcone = matchingDR.at(lowEdge - etaBins.begin() - 1);
+    }
+
+    if (tsos.isValid()) {
+      edm::OwnVector<TrackingRecHit> container;
+
+      if (useOfflineSeed) {
+        // Get the compatible dets on the layer
+        std::vector<pair<const GeomDet *, TrajectoryStateOnSurface>> detsWithStates =
+            detLayer->compatibleDets(tsos, *theService->propagator(thePropagatorName), *theEstimator);
+
+        if (detsWithStates.empty() && barrel) {
+          // Fallback solution using ME2, try again to propagate but using ME2 as reference
+          DetId fallback_id;
+          theta < Geom::pi() / 2. ? fallback_id = CSCDetId(1, 2, 0, 0, 0) : fallback_id = CSCDetId(2, 2, 0, 0, 0);
+          const DetLayer *ME2DetLayer = theService->detLayerGeometry()->idToLayer(fallback_id);
+
+          tsos = theService->propagator(thePropagatorName)->propagate(state, ME2DetLayer->surface());
+          detsWithStates = ME2DetLayer->compatibleDets(tsos, *theService->propagator(thePropagatorName), *theEstimator);
+        }
+
+        if (!detsWithStates.empty()) {
+          TrajectoryStateOnSurface newTSOS = detsWithStates.front().second;
+          const GeomDet *newTSOSDet = detsWithStates.front().first;
+
+          LogDebug(metname) << "Most compatible det";
+          LogDebug(metname) << debug.dumpMuonId(newTSOSDet->geographicalId());
+
+          if (newTSOS.isValid()) {
+            LogDebug(metname) << "pos: (r=" << newTSOS.globalPosition().mag()
+                              << ", phi=" << newTSOS.globalPosition().phi()
+                              << ", eta=" << newTSOS.globalPosition().eta() << ")";
+            LogDebug(metname) << "mom: (q*pt=" << newTSOS.charge() * newTSOS.globalMomentum().perp()
+                              << ", phi=" << newTSOS.globalMomentum().phi()
+                              << ", eta=" << newTSOS.globalMomentum().eta() << ")";
+
+            const TrajectorySeed *assoOffseed =
+                associateOfflineSeedToL1(offlineSeedHandle, offlineSeedMap, newTSOS, dRcone);
```

Comment:
please use a smart pointer.

Reply:
no

Reply:
if there's a bug in `associateOfflineSeedToL1` this can return a null pointer. In other new reco code, we have asked for stricter memory management.